### PR TITLE
Fix slow stats.fisher_exact(x) (#9231)

### PR DESCRIPTION
**Pull request by akahard2dj**
#9231 
It is the PR for the problem that the computation time is long in the situation where x = [[5829225, 5692693], [5760959, 5760959]].

If the value of `pexact` is very small(`np.finfo(np.float64).eps`), the `binary_search()` function is skipped.

### CODE REVIEW

**Comment by ilayn** [comment]
It also needs the test that covers the issue led to this change.

**Comment by akahard2dj** [comment]
The test was conducted before PR.
The scipy.stats test(`scipy.stats.test()`) showed no problems.
Please let me know if there are any additional tests that need to be done.

**Comment by ilayn** [comment]
I mean a new test addition to the test suite that uses the data `x = [[5829225, 5692693], [5760959, 5760959]]`

**Comment by akahard2dj** [comment]
Ah! Thanks!
I apply the case raised in #9231 to the test function.

**Review by ev-br — COMMENTED** [review]
Only a style comment.

File: scipy/stats/stats.py [hunk 1]
```diff
@@ -3157,16 +3157,18 @@ def binary_search(n, n1, n2, side):
...
-            while hypergeom.pmf(guess, n1 + n2, n1, n) > pexact / epsilon:
-                guess += 1
-        else:
-            while hypergeom.pmf(guess, n1 + n2, n1, n) < pexact * epsilon:
-                guess += 1
-            while guess > 0 and hypergeom.pmf(guess, n1 + n2, n1, n) > pexact / epsilon:
-                guess -= 1
+            
+        if not np.isclose(pexact, 0.0, atol=np.finfo(np.float64).eps):
+            if side == "upper":
```
[the remark this answers is not in the document]
Comment (ilayn) [reply scipy/stats/stats.py]: Also `np.spacing(1.0)` might be a bit better for that number.

[the remark this answers is not in the document]
Comment (akahard2dj) [reply scipy/stats/stats.py]: I changed the code style and numpy function(`np.spacing()`).
`scipy.stats.test()` also passed.

**Comment by chrisb83** [comment]
I just had a look at Wikipedia and the following R package: 
https://cran.r-project.org/web/packages/exact2x2/index.html

It seems that the two-sided p-value can be defined in different ways. There is no reference in scipy and I did not look into the code to understand how the calculation works. Any ideas? 
Also, for the test case, are we sure that the results of order 10**(-178) is correct?

**Comment by josef-pkt** [comment]
@chrisb83 
If I understand what you mean:
fisher exact is `minlike` according to my understanding
https://github.com/statsmodels/statsmodels/issues/2735
https://github.com/scipy/scipy/issues/4130#issuecomment-330875018
(those two are cross-linked)

**Comment by akahard2dj** [comment]
@chrisb83 
I actually got the results through `scipy.stats.fisher_exact()`, but the p-value from the example of the issue raised is questionable.

File: scipy/stats/tests/test_stats.py [hunk 2]
```diff
@@ -292,6 +292,8 @@ class TestFisherExact(object):
     def test_basic(self):
         fisher_exact = stats.fisher_exact
 
+        res = fisher_exact([[5829225, 5692693], [5760959, 5760959]])[1]
+        assert_approx_equal(res, 3.06311e-178, significant=4)
```
Comment (sethtroisi) [inline scipy/stats/tests/test_stats.py:296]: I asked R for the value and got 
```
> fisher.test(matrix(c(5829225, 5692693, 5760959, 5760959), nrow=2, ncol=2))[1]
$p.value
[1] 6.126213e-178
```

Which is not significant = 4

Comment (WarrenWeckesser) [reply scipy/stats/tests/test_stats.py:296]: Looks like the value `3.06311e-178` is wrong.  I did an independent calculation (a slow, brute force method using the Gnu GMP library) and got a p-value of `6.12621271262e-178`.  That matches the R result up to the number of digits that I recorded:

```
> options(digits=12)
> result = fisher.test(matrix(c(5829225, 5692693, 5760959, 5760959), nrow=2, ncol=2))
> result$p.value
[1] 6.12621271262e-178
```

Comment (WarrenWeckesser) [reply scipy/stats/tests/test_stats.py:296]: Hold on, though.  I also calculated the equivalent of `alternative='greater'` and that value is `3.06310635631e-178`, which also agrees with R:
```
> result = fisher.test(matrix(c(5829225, 5692693, 5760959, 5760959), nrow=2, ncol=2), alternative='greater')
> result$p.value
[1] 3.06310635631e-178
```
Just noting that fact for future reviewers.  I haven't looked into why the above test ended up with what appears to be the result for the case `alternative='greater'`.

**Comment by mdhaber** [comment]
@rgommers I just got notifications about the closure of a lot of issues. I assume it had something to do with deleting `master`? Should these be re-opened if there is not evidence that the issue is closed?

**Comment by mdhaber** [comment]
Ooops maybe it was because the repo that submitted the PR was deleted.

**Comment by rgommers** [comment]
> Ooops maybe it was because the repo that submitted the PR was deleted. The underlying issue is still open, so it's fine.

Yes, that should be the reason. Otherwise PRs should just be pointed at `main` by GitHub automatically.

**Comment by rgommers** [comment]
Commits are still present, so if there's PRs of value, it'd be good to either make a note on the relevant issue, or take them over and resubmit them.

**Pull request closed without merging**